### DISCUSSION TITLE
Enhancement Request:  Create console command for adjusting ADC multi-sampling settings.

### DISCUSSION THREAD

**Discussion by LucidEye** [arendst/Tasmota#17974] [Ideas]
Hello,

I am using several analog sensors for various agricultural monitors and data loggers.  I am noticing that the data signal from the ESP32's built-in ADC is very noisy and has a lot of "jitter" in the returned data.  I am aware of the issues and limitations with the built-in ADC, not truly linear, and quite noisy without filtering capacitors, but Espressif's own documentation recommends that BOTH the filtering capacitors AND multi-sampling be used whenever possible to get more "accurate" readings from the ADC and reduce the noise and jitter.
https://docs.espressif.com/projects/esp-idf/en/v4.3.3/esp32/api-reference/peripherals/adc.html#minimizing-noise

I was told in the Discord discussion that Tasmota does have multi-sampling implemented and enabled, however it is by default only set to 2-samples per reading. This setting can be found in the following file, lines 314-328:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_02_analog.ino

`uint16_t AdcRead(uint32_t pin, uint32_t factor) {
  // factor 1 = 2 samples
  // factor 2 = 4 samples
  // factor 3 = 8 samples
  // factor 4 = 16 samples
  // factor 5 = 32 samples
  uint32_t samples = 1 << factor;
  uint32_t analog = 0;
  for (uint32_t i = 0; i < samples; i++) {
    analog += analogRead(pin);
    delay(1);
  }
  analog >>= factor;
  return analog;
}`

I was also informed that to change this setting requires an entire re-compile of the firmware?  This really seems like a feature that is important enough that it should have a console command to expose this setting for easier adjustment.  Is there such a command? If not I would like to request this as a feature enhancement if possible.

Thank you for any help or info on this.

**Comment by barbudor** [comment]
You pointed the right lines where `factor` defines the oversampling
But did you tried to look further to see value `factor` was set to ?

If you are using ADC_INPUT, the factor is set to 5 which means 32 readings
Not enough ?

**Reply by LucidEye** [reply]
I'm sorry, I did not see that, I thought the code I posted showed that the sampling is set to "1" which is only 2-samples per measurement.  Can you please post where this setting is in the code for ANALOG_RANGE?  Also, is this setting specific to the type of ADC sensor (1-8) or is the multi-sampling setting global for all of the ADC channels and sensor types?  I only use the ADC_INPUT for calibrating with the raw values, then I switch to the ADC_RANGE for the actual readings once I calibrate using the AdcParam 6 command.  If this is the correct part of the code (460-469), it looks to me like ADC_RANGE is set to a factor of 1?  Again, I am not a programmer, so trying to figure this out is a bit of a learning curve for me at the moment.... please bear with me, and thank you for the help.

`void AdcGetCurrentPower(uint8_t idx, uint8_t factor) {
  // factor 1 = 2 samples
  // factor 2 = 4 samples
  // factor 3 = 8 samples
  // factor 4 = 16 samples
  // factor 5 = 32 samples
  uint8_t samples = 1 << factor;
  uint16_t analog = 0;
  uint16_t analog_min = ANALOG_RANGE;
  uint16_t analog_max = 0;

`

**Reply by barbudor** [reply]
So ADC_RANGE use a factor of 2 :(

https://github.com/arendst/Tasmota/blob/d8d9c549ab62803b251e5a09aaf43848c1f7c97b/tasmota/tasmota_xsns_sensor/xsns_02_analog.ino#L451-L458

Doesn't make much sense to use a different value
I will push for 5 too

**Reply by barbudor** [reply]
https://github.com/arendst/Tasmota/pull/17975

**Reply by LucidEye** [reply]
Thank you.  Yeah, it doesn't make much sense that the defaults are set so low when Espressif's documentation says to do the opposite  I would like to add the 0.1 uf filter caps, but it would be kind of pointless if the multi-sampling isn't also set to at least factor 4. :-/
It would still be nice to be able to change these settings from the Console if possible for troubleshooting and calibration purposes.

**Reply by barbudor** [reply]
It has been merged
It's now availale in latest development builds from http://ota.tasmota.com/tasmota/, and soon for ESP32 at http://ota.tasmota.com/tasmota32/

Let us know if it improve
We'll see if making configurable drag some interresy

**Reply by LucidEye** [reply]
Wow!  That was fast!  Thank you!  So I should backup/export all my current settings from that device, and then I can download the latest build off that page (tasmota-4M.bin), flash it to my esp32, and then import my saved settings and it should then be using the 32x multi-sampling on the ADC_RANGE option?

**Reply by barbudor** [reply]
You can do OTA update directly
Go to the upgrade firmware page from the main menu and in the 1st entry field enter url
http://ota.tasmota.com/tasmota/tasmota-4M.bin.gz
Then click on the upgrade button

Tasmota Wil automatically upgrade first to Tasmota-minimal intermediate step and then to tasmota-4m

It's always a good idea to save the settings just in case but normally that should be seamless

**Reply by barbudor** [reply]
Stop
tasmota-4m is not for esp32

What exact device do you have and which binary are you running?

**Reply by LucidEye** [reply]
![Screenshot at 2023-02-17 15-54-22](https://user-images.githubusercontent.com/12551280/219812685-6bdc0024-2259-4c4b-8d8f-560be2d64b91.png)
![Screenshot at 2023-02-17 15-54-03](https://user-images.githubusercontent.com/12551280/219812686-4f887f0f-5d6f-4154-8939-e1de9c83e44f.png)

https://www.amazon.com/dp/B09Z7Q5LKQ?psc=1&ref=ppx_yo2ov_dt_b_product_details

**Reply by barbudor** [reply]
OK, so you need
http://ota.tasmota.com/tasmota32/tasmota32.bin

Same process but different url

**Reply by LucidEye** [reply]
Thanks again!  Glad I caught your reply before flashing :-P

**Reply by LucidEye** [reply]
So just a quick follow-up. I still do not have the capacitors installed yet, but just with the multi-sampling set higher the signal is already much "cleaner"... less bounce and jitter. 
Everything on the left of 17:00 is with the 2x multi-sampling, everything on the right is with the 32x multi-sampling...
![Screenshot at 2023-02-17 18-00-18](https://user-images.githubusercontent.com/12551280/219823393-8a914b31-c510-4f78-98d7-3bd292906c74.png)

So once I get the filter caps installed it should be even more of an improvement as the caps should attenuate most of the high-reading spikes.  

Thank you so much once again :-)
